rdm/render.py

```diff
@@ -32,30 +33,73 @@ def join_to(foreign_keys, table, primary_key='id'):
     return joined
 
 
-def render_template(template_filename, context, output_file):
-    loader = jinja2.ChoiceLoader([
-        jinja2.FileSystemLoader('.'),
-        jinja2.PackageLoader('rdm', '.'),
-    ])
+def render_template_to_file(template_filename, context, output_file, loaders=None):
+    generator = generate_template_output(template_filename, context, loaders=loaders)
+    TemplateStream(generator).dump(output_file)
+
+
+def render_template_to_string(template_filename, context, loaders=None):
+    return ''.join(generate_template_output(template_filename, context, loaders=loaders))
+
+
+def generate_template_output(template_filename, context, loaders=None):
+    environment = _create_jinja_environment(context, loaders)
+    template = environment.get_template(template_filename)
+    source_line_list = _generate_source_line_list(template, context)
+    return _generate_output_lines(environment, source_line_list)
+
+
+def _create_jinja_environment(context, loaders=None):
+    extensions = _create_extension_list(context)
+    loader = _create_loader(loaders)
     environment = jinja2.Environment(
+        cache_size=0,
         undefined=jinja2.StrictUndefined,
         loader=loader,
-        extensions=[AuditNoteExtension],
+        extensions=extensions,
     )
-
     environment.filters['invert_dependencies'] = invert_dependencies
     environment.filters['join_to'] = join_to
+    return environment
 
+
+def _create_extension_list(context):
     system_dict = context.get('system', {})
-    audit_notes = system_dict.get('auditor_notes')
-    if audit_notes:
-        environment.audit_note_default_formatter = plain_formatter
-        special_formats = system_dict.get('auditor_note_formats')
-        if special_formats:
-            for format_tag, formatter in special_formats.items():
-                if isinstance(formatter, str):
-                    formatter = create_formatter_with_string(formatter)
-                environment.audit_note_formatting_dictionary[format_tag] = formatter
+    extension_descriptor_list = system_dict.get('extension_load_list', [])
+    return dynamic_class_loader(extension_descriptor_list)
 
-    template = environment.get_template(template_filename)
-    template.stream(**context).dump(output_file)
+
+def _create_loader(loaders=None):
+    if loaders is None:
+        loaders = [
+            jinja2.FileSystemLoader('.'),
+            jinja2.PackageLoader('rdm', '.'),
+        ]
+
+    return jinja2.ChoiceLoader(loaders)
+
+
+def _generate_source_line_list(template, context):
+    generator = template.generate(**context)
+    for filter in [split_into_lines, append_newlines]:
```

Comment:
Can we assume that `generator` will not produces strings that strattle new lines?  E.g., what if it returned:

1. `first line\nsecond'
2. ' line\nthird line'

It seems like the final output would be a generator that produces these strings:

```
['first line\n', 'second\n', 'line\n', 'third line\n']
```

Reply:
This needs to be looked at quite carefully.